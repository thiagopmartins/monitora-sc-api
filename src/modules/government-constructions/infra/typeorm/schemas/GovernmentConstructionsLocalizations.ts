import {
  Column,
  CreateDateColumn,
  Entity,
  Index,
  ObjectID,
  ObjectIdColumn,
  UpdateDateColumn,
} from 'typeorm';

@Entity('government-constructions-localizations')
class GovernmentConstructionsLocalizations {
  @ObjectIdColumn()
  id: ObjectID;

  @Column()
  @Index({ unique: true })
  construction_id: string;

  @Column()
  description: string;

  @Column()
  nuLatitude: number;

  @Column()
  nuLongitude: number;

  @Column()
  urlPin: string;

  @Column()
  heatMapValue: number;

  @Column()
  coordenates: number[];

  @Column()
  projectsId: number[];

  @Column()
  contracts: {
    id: number;
    constructionSitesId: number[];
  }[];

  @Column()
  heatMapQuantity: number;

  @CreateDateColumn()
  created_at: Date;

  @UpdateDateColumn()
  updated_at: Date;
}

export default GovernmentConstructionsLocalizations;
