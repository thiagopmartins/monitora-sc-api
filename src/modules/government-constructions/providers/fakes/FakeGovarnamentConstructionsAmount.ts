import GovernmentConstructionsAmounts from '@modules/government-constructions/infra/typeorm/schemas/GovernmentConstructionsAmounts';
import AppError from '@shared/errors/AppError';
import IGovernmentConstructionAmountProvider from '../IGovernmentConstructionAmountProvider';

class FakeGovarnamentConstructionsAmount
  implements IGovernmentConstructionAmountProvider {
  private constructions: GovernmentConstructionsAmounts[] = [];

  public async getAll(): Promise<GovernmentConstructionsAmounts[] | undefined> {
    try {
      const construction = {
        id: 7192,
        number: '624',
        year: 2020,
        description: 'CONSTRUÇÃO DA MURALHA - CASEP CONCÓRDIA ',
        area: 'Justiça e Cidadania',
        status: 'A iniciar',
        amount: 469780.47,
        cityId: 8083,
        cityDescription: 'Concórdia',
        projectLink: null,
        peExecuted: 0.0,
        cities: [],
        contracts: [
          {
            id: 1031437,
            description: null,
            number: null,
          },
        ],
        constructionSites: [],
      };

      const governamentConstructionsAmounts = new GovernmentConstructionsAmounts();

      Object.assign(
        governamentConstructionsAmounts,
        { construction_amount_id: construction.id },
        construction,
      );

      delete governamentConstructionsAmounts.id;

      this.constructions.push(governamentConstructionsAmounts);

      return this.constructions;
    } catch (error) {
      throw new AppError(error);
    }
  }
}

export default FakeGovarnamentConstructionsAmount;
