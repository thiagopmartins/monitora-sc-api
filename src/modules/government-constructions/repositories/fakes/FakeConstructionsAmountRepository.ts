import ICreateConstructionsAmountDTO from '@modules/government-constructions/dtos/ICreateConstructionsAmountDTO';
import IFindConstructionsAmountDTO from '@modules/government-constructions/dtos/IFindConstructionsAmountDTO';
import GovernmentConstructionsAmounts from '@modules/government-constructions/infra/typeorm/schemas/GovernmentConstructionsAmounts';
import { uuid } from 'uuidv4';
import IGovernmentConstructionsAmountsRepository from '../IGovernmentConstructionsAmountsRepository';

class FakeConstructionsAmountRepository
  implements IGovernmentConstructionsAmountsRepository {
  public constructions: GovernmentConstructionsAmounts[] = [];

  public async create(
    constructionDTO: ICreateConstructionsAmountDTO,
  ): Promise<GovernmentConstructionsAmounts> {
    const construction = new GovernmentConstructionsAmounts();

    const date = new Date();

    Object.assign(
      construction,
      { id: uuid(), created_at: date, updated_at: date },
      constructionDTO,
    );
    this.constructions.push(construction);

    return construction;
  }

  public async update(
    construction: GovernmentConstructionsAmounts,
  ): Promise<GovernmentConstructionsAmounts> {
    const findIndex = this.constructions.findIndex(
      findConstruction => findConstruction.id === construction.id,
    );

    this.constructions[findIndex] = construction;

    return construction;
  }

  public async findByConstructionAmountId(
    id: number,
  ): Promise<GovernmentConstructionsAmounts | undefined> {
    const findConstruction = this.constructions.find(
      construction => construction.construction_amount_id === id,
    );
    return findConstruction;
  }

  public async find(
    filter: IFindConstructionsAmountDTO | undefined,
  ): Promise<GovernmentConstructionsAmounts[]> {
    if (filter === undefined) {
      return this.constructions;
    }

    if (filter.year === undefined) {
      return this.constructions.filter(x => filter.status.includes(x.status));
    }

    if (filter.status === undefined) {
      return this.constructions.filter(x => {
        return x.year >= filter.year.initial && x.year <= filter.year.finish;
      });
    }

    return this.constructions.filter(
      x =>
        filter.status.includes(x.status) &&
        x.year >= filter.year.initial &&
        x.year <= filter.year.finish,
    );
  }
}

export default FakeConstructionsAmountRepository;
