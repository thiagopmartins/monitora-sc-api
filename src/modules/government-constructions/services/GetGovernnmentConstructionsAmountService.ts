import { inject, injectable } from 'tsyringe';
import ICreateConstructionsAmountDTO from '../dtos/ICreateConstructionsAmountDTO';
import IGovernmentConstructionAmountProvider from '../providers/IGovernmentConstructionAmountProvider';
import IGovernmentConstructionsAmountsRepository from '../repositories/IGovernmentConstructionsAmountsRepository';

@injectable()
class GetGovernnmentConstructionsAmountService {
  constructor(
    @inject('GovernmentConstructionsAmountsRepository')
    private governmentConstructionsAmountsRepository: IGovernmentConstructionsAmountsRepository,

    @inject('GovernmentConstructionsAmountsProvider')
    private governmentConstructionsAmountsProvider: IGovernmentConstructionAmountProvider,
  ) {}

  public async execute(): Promise<void> {
    const result = await this.governmentConstructionsAmountsProvider.getAll();
    if (result !== undefined) {
      result.map(async construction => {
        await this.governmentConstructionsAmountsRepository.create(
          construction as ICreateConstructionsAmountDTO,
        );
      });
    }
  }
}

export default GetGovernnmentConstructionsAmountService;
