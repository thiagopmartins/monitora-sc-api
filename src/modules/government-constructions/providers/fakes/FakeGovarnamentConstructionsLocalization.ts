import GovernmentConstructionsLocalizations from '@modules/government-constructions/infra/typeorm/schemas/GovernmentConstructionsLocalizations';
import AppError from '@shared/errors/AppError';
import IGovernmentConstructionLocalizationProvider from '../IGovernmentConstructionLocalizationProvider';

class FakeGovarnamentConstructionsLocalization
  implements IGovernmentConstructionLocalizationProvider {
  private constructions: GovernmentConstructionsLocalizations[] = [];

  public async getAll(): Promise<
    GovernmentConstructionsLocalizations[] | undefined
  > {
    try {
      const construction = {
        id: 'SC110 _24',
        description:
          'PETROLÃ‚NDIA (ENTR. SC-284) - ENTR. BR-282 (P/ BOCAINA DO SUL)',
        nuLatitude: -27.535722175999,
        nuLongitude: -49.698887383295,
        urlPin:
          'http://www.sicop.sc.gov.br:80/mapa-smo-backend/imagens/areas/345pin.png',
        heatMapValue: 254912.79,
        coordenates: [],
        projectsId: [2239, 2259, 3992, 7192],
        contracts: [
          {
            id: 1026464,
            constructionSitesId: [1],
          },
          {
            id: 1027889,
            constructionSitesId: [1],
          },
          {
            id: 1029711,
            constructionSitesId: [1],
          },
        ],
        heatMapQuantity: 2.12,
      };

      const governamentConstructionsLocalizations = new GovernmentConstructionsLocalizations();

      Object.assign(
        governamentConstructionsLocalizations,
        { construction_id: construction.id },
        construction,
      );

      delete governamentConstructionsLocalizations.id;

      this.constructions.push(governamentConstructionsLocalizations);

      return this.constructions;
    } catch (error) {
      throw new AppError(error);
    }
  }
}

export default FakeGovarnamentConstructionsLocalization;
