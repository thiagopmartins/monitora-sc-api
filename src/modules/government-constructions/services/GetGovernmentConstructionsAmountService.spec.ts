import GovernmentConstructionsLocalizations from '../infra/typeorm/schemas/GovernmentConstructionsLocalizations';
import FakeGovarnamentConstructionsAmount from '../providers/fakes/FakeGovarnamentConstructionsAmount';
import FakeGovarnamentConstructionsLocalization from '../providers/fakes/FakeGovarnamentConstructionsLocalization';
import FakeConstructionsAmountRepository from '../repositories/fakes/FakeConstructionsAmountRepository';
import GetGovernnmentConstructionsAmountService from './GetGovernmentConstructionsAmountService';

describe('GetGovernnmentConstructionsAmountService', () => {
  let fakeConstructionsAmountRepository: FakeConstructionsAmountRepository;
  let fakeGovarnamentConstructionsAmount: FakeGovarnamentConstructionsAmount;
  let getGovernmentConstructionsAmount: GetGovernnmentConstructionsAmountService;
  let fakeGovarnamentConstructionsLocalization: FakeGovarnamentConstructionsLocalization;
  let governmentConstructionsLocalizationList:
    | GovernmentConstructionsLocalizations[]
    | undefined;

  beforeAll(async () => {
    fakeConstructionsAmountRepository = new FakeConstructionsAmountRepository();
    fakeGovarnamentConstructionsAmount = new FakeGovarnamentConstructionsAmount();
    fakeGovarnamentConstructionsLocalization = new FakeGovarnamentConstructionsLocalization();
    getGovernmentConstructionsAmount = new GetGovernnmentConstructionsAmountService(
      fakeConstructionsAmountRepository,
      fakeGovarnamentConstructionsAmount,
    );

    governmentConstructionsLocalizationList = await fakeGovarnamentConstructionsLocalization.getAll();
  });

  it('should be able to create a new construction amount', async () => {
    const create = jest.spyOn(fakeConstructionsAmountRepository, 'create');

    await getGovernmentConstructionsAmount.execute(
      governmentConstructionsLocalizationList,
    );

    expect(create).toBeCalledWith(
      expect.objectContaining({
        construction_amount_id: expect.any(Number),
        latitude: expect.any(Number),
        longitude: expect.any(Number),
      }),
    );

    expect(fakeConstructionsAmountRepository.constructions).toMatchSnapshot([
      {
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
        construction_amount_id: expect.any(Number),
        id: expect.any(String),
      },
    ]);
  });

  it('should be able to update if already exists construcion', async () => {
    const update = jest.spyOn(fakeConstructionsAmountRepository, 'update');

    await getGovernmentConstructionsAmount.execute(
      governmentConstructionsLocalizationList,
    );
    await getGovernmentConstructionsAmount.execute(
      governmentConstructionsLocalizationList,
    );

    expect(update).toBeCalledWith(
      expect.objectContaining({
        construction_amount_id: expect.any(Number),
      }),
    );

    expect(fakeConstructionsAmountRepository.constructions).toMatchSnapshot([
      {
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
        construction_amount_id: expect.any(Number),
        id: expect.any(String),
      },
    ]);
  });
  it('should be not call repository functions if governmentConstructionsAmountProvider is undefined', async () => {
    jest
      .spyOn(fakeGovarnamentConstructionsAmount, 'getAll')
      .mockImplementation(() => Promise.resolve(undefined));

    const find = jest.spyOn(
      fakeConstructionsAmountRepository,
      'findByConstructionAmountId',
    );

    await getGovernmentConstructionsAmount.execute(
      governmentConstructionsLocalizationList,
    );

    expect(find).not.toBeCalled();
  });
});
