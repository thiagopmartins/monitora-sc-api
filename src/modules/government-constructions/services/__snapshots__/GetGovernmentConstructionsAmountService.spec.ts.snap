// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetGovernnmentConstructionsAmountService should be able to create a new construction amount 1`] = `
Array [
  Object {
    "amount": 469780.47,
    "area": "Justiça e Cidadania",
    "cities": Array [],
    "cityDescription": "Concórdia",
    "cityId": 8083,
    "constructionSites": Array [],
    "construction_amount_id": Any<Number>,
    "contracts": Array [
      Object {
        "description": null,
        "id": 1031437,
        "number": null,
      },
    ],
    "created_at": Any<Date>,
    "description": "CONSTRUÇÃO DA MURALHA - CASEP CONCÓRDIA ",
    "id": Any<String>,
    "latitude": -27.535722175999,
    "longitude": -49.698887383295,
    "number": "624",
    "peExecuted": 0,
    "projectLink": null,
    "status": "A iniciar",
    "updated_at": Any<Date>,
    "year": 2020,
  },
]
`;

exports[`GetGovernnmentConstructionsAmountService should be able to update if already exists construcion 1`] = `
Array [
  Object {
    "amount": 469780.47,
    "area": "Justiça e Cidadania",
    "cities": Array [],
    "cityDescription": "Concórdia",
    "cityId": 8083,
    "constructionSites": Array [],
    "construction_amount_id": Any<Number>,
    "contracts": Array [
      Object {
        "description": null,
        "id": 1031437,
        "number": null,
      },
    ],
    "created_at": Any<Date>,
    "description": "CONSTRUÇÃO DA MURALHA - CASEP CONCÓRDIA ",
    "id": Any<String>,
    "latitude": -27.535722175999,
    "longitude": -49.698887383295,
    "number": "624",
    "peExecuted": 0,
    "projectLink": null,
    "status": "A iniciar",
    "updated_at": Any<Date>,
    "year": 2020,
  },
]
`;
